File: android/src/main/java/ru/vvdev/yamap/view/YamapMarker.java
package ru.vvdev.yamap.view;

import android.animation.ValueAnimator;
import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.PointF;
import android.os.Looper;
import android.view.View;
import android.view.animation.LinearInterpolator;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.uimanager.events.RCTEventEmitter;
import com.facebook.react.views.view.ReactViewGroup;
import com.yandex.mapkit.geometry.Point;
import com.yandex.mapkit.map.IconStyle;
import com.yandex.mapkit.map.MapObject;
import com.yandex.mapkit.map.MapObjectTapListener;
import com.yandex.mapkit.map.PlacemarkMapObject;
import com.yandex.mapkit.map.RotationType;
import com.yandex.runtime.image.ImageProvider;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.Timer;
import java.util.TimerTask;
import java.util.logging.Handler;

import ru.vvdev.yamap.models.ReactMapObject;
import ru.vvdev.yamap.utils.Callback;
import ru.vvdev.yamap.utils.ImageLoader;

public class YamapMarker extends ReactViewGroup implements MapObjectTapListener, ReactMapObject {
    public Point point;
    private int zIndex = 1;
    private float scale = 1;
    private Boolean visible = true;
    private Boolean rotated = false;
    private int YAMAP_FRAMES_PER_SECOND = 25;
    private PointF markerAnchor = null;
    private String iconSource;
    private View _childView;
    private PlacemarkMapObject mapObject;
    private ArrayList<View> childs = new ArrayList<>();

    private OnLayoutChangeListener childLayoutListener = new OnLayoutChangeListener() {
        @Override
        public void onLayoutChange(View v, int left, int top, int right, int bottom, int oldLeft, int oldTop, int oldRight, int oldBottom) {
            updateMarker();
        }
    };

    public YamapMarker(Context context) {
        super(context);
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
    }

    // PROPS
    public void setPoint(Point _point) {
        point = _point;
        updateMarker();
    }

    public void setZIndex(int _zIndex) {
        zIndex = _zIndex;
        updateMarker();
    }

    public void setScale(float _scale) {
        scale = _scale;
        updateMarker();
    }

    public void setRotated(Boolean _rotated) {
        rotated = _rotated;
        updateMarker();
    }

    public void setVisible(Boolean _visible) {
        visible = _visible;
        updateMarker();
    }

    public void setIconSource(String source) {
        iconSource = source;
        updateMarker();
    }

    public void setAnchor(PointF anchor) {
        markerAnchor = anchor;
        updateMarker();
    }

    private void updateMarker() {
        if (mapObject != null && mapObject.isValid()) {
            final IconStyle iconStyle = new IconStyle();
            iconStyle.setScale(scale);
            iconStyle.setRotationType(rotated ? RotationType.ROTATE : RotationType.NO_ROTATION);
            iconStyle.setVisible(visible);
            if (markerAnchor != null) {
                iconStyle.setAnchor(markerAnchor);
            }
            mapObject.setGeometry(point);
            mapObject.setZIndex(zIndex);
            mapObject.setIconStyle(iconStyle);

            if (_childView != null) {
                try {
                    Bitmap b = Bitmap.createBitmap(_childView.getWidth(), _childView.getHeight(), Bitmap.Config.ARGB_8888);
                    Canvas c = new Canvas(b);
                    _childView.draw(c);
                    mapObject.setIcon(ImageProvider.fromBitmap(b));
                    mapObject.setIconStyle(iconStyle);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
            if (childs.size() == 0) {
                if (!iconSource.equals("")) {
                    YamapView parent = (YamapView)getParent();
                    if (parent!=null) {
                        parent.setImage(iconSource, mapObject, iconStyle);
                    }
                }
            }
        }
    }

    public void setMapObject(MapObject obj) {
        mapObject = (PlacemarkMapObject) obj;
        mapObject.addTapListener(this);
        updateMarker();
    }

    public MapObject getMapObject() {
        return mapObject;
    }

    public void setChildView(View view) {
        if (view == null) {
            _childView.removeOnLayoutChangeListener(childLayoutListener);
            _childView = null;
            updateMarker();
            return;
        }
        _childView = view;
        _childView.addOnLayoutChangeListener(childLayoutListener);
    }

    public void addChildView(View view, int index) {
        childs.add(index, view);
        setChildView(childs.get(0));
    }

    public void removeChildView(int index) {
        childs.remove(index);
        setChildView(childs.size() > 0 ? childs.get(0) : null);
    }

    public void moveAnimationLoop(double lat, double lon) {
        PlacemarkMapObject placemark = (PlacemarkMapObject) this.getMapObject();
        placemark.setGeometry(new Point(lat, lon));
    }

    public void rotateAnimationLoop(float delta) {
        PlacemarkMapObject placemark = (PlacemarkMapObject) this.getMapObject();
        placemark.setDirection(delta);
    }

    public void animatedMoveTo(Point point, final float duration) {
        PlacemarkMapObject placemark = (PlacemarkMapObject) this.getMapObject();
        Point p = placemark.getGeometry();
        final double startLat = p.getLatitude();
        final double startLon = p.getLongitude();
        final double deltaLat = point.getLatitude() - startLat;
        final double deltaLon = point.getLongitude() - startLon;
        ValueAnimator valueAnimator = ValueAnimator.ofFloat(0, 1);
        valueAnimator.setDuration((long) duration);
        valueAnimator.setInterpolator(new LinearInterpolator());
        valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override public void onAnimationUpdate(ValueAnimator animation) {
                try {
                    float v = animation.getAnimatedFraction();
                    moveAnimationLoop(startLat + v*deltaLat, startLon + v*deltaLon);
                } catch (Exception ex) {
                    // I don't care atm..
                }
            }
        });
        valueAnimator.start();
    }

    public void animatedRotateTo(final float angle, float duration) {
        PlacemarkMapObject placemark = (PlacemarkMapObject) this.getMapObject();
        final float startDirection = placemark.getDirection();
        final float delta = angle - placemark.getDirection();
        ValueAnimator valueAnimator = ValueAnimator.ofFloat(0, 1);
        valueAnimator.setDuration((long) duration);
        valueAnimator.setInterpolator(new LinearInterpolator());
        valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override public void onAnimationUpdate(ValueAnimator animation) {
                try {
                    float v = animation.getAnimatedFraction();
                    rotateAnimationLoop(startDirection + v*delta);
                } catch (Exception ex) {
                    // I don't care atm..
                }
            }
        });
        valueAnimator.start();
    }

    @Override
    public boolean onMapObjectTap(@NonNull MapObject mapObject, @NonNull Point point) {
        WritableMap e = Arguments.createMap();
        ((ReactContext) getContext()).getJSModule(RCTEventEmitter.class).receiveEvent(getId(), "onPress", e);

        return false;
    }
}

File: android/src/main/java/ru/vvdev/yamap/view/YamapMarker.kt
package ru.vvdev.yamap.view

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PointF
import android.graphics.RectF
import android.view.View
import android.view.View.OnLayoutChangeListener
import android.view.animation.LinearInterpolator
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactContext
import com.facebook.react.uimanager.events.RCTEventEmitter
import com.facebook.react.views.view.ReactViewGroup
import com.yandex.mapkit.geometry.Point
import com.yandex.mapkit.map.IconStyle
import com.yandex.mapkit.map.MapObject
import com.yandex.mapkit.map.MapObjectTapListener
import com.yandex.mapkit.map.PlacemarkMapObject
import com.yandex.mapkit.map.RotationType
import com.yandex.runtime.image.ImageProvider
import ru.vvdev.yamap.models.ReactMapObject
import ru.vvdev.yamap.utils.Callback
import ru.vvdev.yamap.utils.ImageLoader.DownloadImageBitmap
import android.graphics.Color
import android.graphics.Paint.ANTI_ALIAS_FLAG
import android.graphics.Path
import android.graphics.Shader
import android.graphics.BitmapShader
import android.graphics.BitmapFactory


class YamapMarker(context: Context?) : ReactViewGroup(context), MapObjectTapListener,
    ReactMapObject {
    @JvmField
    var point: Point? = null
    private var zIndex = 1
    private var scale = 1f
    private var visible = true
    private var handled = true
    private var rotated = false
    private val YAMAP_FRAMES_PER_SECOND = 25
    private var markerAnchor: PointF? = null
    private var iconSource: String? = null
    private var _childView: View? = null
    override var rnMapObject: MapObject? = null
    private val childs = ArrayList<View>()
    private var similarMarkersCount: Int? = null

    private val childLayoutListener =
        OnLayoutChangeListener { v, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom -> updateMarker() }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
    }

    // PROPS
    fun setPoint(_point: Point?) {
        point = _point
        updateMarker()
    }

    fun setZIndex(_zIndex: Int) {
        zIndex = _zIndex
        updateMarker()
    }

    fun setScale(_scale: Float) {
        scale = _scale
        updateMarker()
    }

    fun setHandled(_handled: Boolean) {
        handled = _handled
    }

    fun setRotated(_rotated: Boolean) {
        rotated = _rotated
        updateMarker()
    }

    fun setVisible(_visible: Boolean) {
        visible = _visible
        updateMarker()
    }

    fun setIconSource(source: String?) {
        iconSource = source
        updateMarker()
    }

    fun setAnchor(anchor: PointF?) {
        markerAnchor = anchor
        updateMarker()
    }

    fun setSimilarMarkersCount(count: Int?) {
        similarMarkersCount = count
        updateMarker()
    }
    
    private fun updateMarker() {
        if (rnMapObject != null && rnMapObject!!.isValid) {
            val iconStyle = IconStyle()
            iconStyle.setScale(scale)
            iconStyle.setRotationType(if (rotated) RotationType.ROTATE else RotationType.NO_ROTATION)
            iconStyle.setVisible(visible)
            if (markerAnchor != null) {
                iconStyle.setAnchor(markerAnchor)
            }
            (rnMapObject as PlacemarkMapObject).geometry = point!!
            (rnMapObject as PlacemarkMapObject).zIndex = zIndex.toFloat()
            (rnMapObject as PlacemarkMapObject).setIconStyle(iconStyle)
        // Проверяем, нужно ли отрисовывать текст
        if (similarMarkersCount != null && similarMarkersCount!! > 1) {
            drawTextMarker(similarMarkersCount!!)
            } else { // если нет, то текущая логика
                if (_childView != null) {
                    try {
                        val b = Bitmap.createBitmap(
                            _childView!!.width, _childView!!.height, Bitmap.Config.ARGB_8888
                        )
                        val c = Canvas(b)
                        _childView!!.draw(c)
                        (rnMapObject as PlacemarkMapObject).setIcon(ImageProvider.fromBitmap(b))
                        (rnMapObject as PlacemarkMapObject).setIconStyle(iconStyle)
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }
                if (childs.size == 0) {
                    if (iconSource != "") { // если есть ссылка на ресурс отрисовывем картинку
                        iconSource?.let {
                            DownloadImageBitmap(context, it, object : Callback<Bitmap?> {
                                override fun invoke(arg: Bitmap?) {
                                    try {
                                        if (arg != null) {
                                            val scaledBitmap = Bitmap.createScaledBitmap(arg, 140, 140, true)
                                            val icon = createRoundedMarkerImage(scaledBitmap)
                                            // val icon = createRoundedMarkerImage(arg)
                                            (rnMapObject as PlacemarkMapObject).setIcon(ImageProvider.fromBitmap(icon))
                                            (rnMapObject as PlacemarkMapObject).setIconStyle(iconStyle)
                                        }
                                    } catch (e: Exception) {
                                        e.printStackTrace()
                                    }
                                }
                            })
                        }
                    }
                }
            }
        }
    }
    // Создаем Bitmap для текстового маркера
    private fun drawTextMarker(count: Int) {
        val textBitmap = Bitmap.createBitmap(140, 140, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(textBitmap)
    
        // Рисуем белый фон с закругленными углами
        val backgroundPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.WHITE
            style = Paint.Style.FILL
        }
        val radius = 80f // Радиус закругления
        canvas.drawRoundRect(RectF(0f, 0f, textBitmap.width.toFloat(), textBitmap.height.toFloat()), radius, radius, backgroundPaint)
    
        // Настраиваем Paint для текста
        val textPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.BLACK
            textSize = 60f
            textAlign = Paint.Align.CENTER
        }
    
        // Отрисовываем текст в центре
        canvas.drawText(count.toString(), (textBitmap.width / 2).toFloat(), (textBitmap.height / 2).toFloat() - (textPaint.descent() + textPaint.ascent()) / 2, textPaint)
    
        // Рисуем черную обводку вокруг закругленного фона
        val borderPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.BLACK
            style = Paint.Style.STROKE
            strokeWidth = 10f // Толщина обводки
        }

        // Рисуем обводку вокруг закругленного фона
        canvas.drawRoundRect(RectF(10f, 10f, (textBitmap.width - 10f).toFloat(), (textBitmap.height - 10f).toFloat()), radius, radius, borderPaint)
    
        // Устанавливаем иконку маркера как текстовый Bitmap
        (rnMapObject as PlacemarkMapObject).setIcon(ImageProvider.fromBitmap(textBitmap))
    }

    // Создаем Bitmap для изображения
    private fun createRoundedMarkerImage(markerImage: Bitmap?): Bitmap? {
        if (markerImage == null) return null
    
        val radius = 80f // Радиус закругления
        val borderWidth = 10f // Ширина границы
        val borderColor = Color.WHITE // Цвет границы
    
        // Создаем BitmapShader из изображения маркера
        val shader = BitmapShader(markerImage, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP)
        val paint = Paint(ANTI_ALIAS_FLAG)
        paint.shader = shader // Устанавливаем Shader для Paint
    
        // Создаем новый Bitmap для закругленного изображения
        val roundedBitmap = Bitmap.createBitmap(
            140 + (borderWidth * 2).toInt(), // Ширина с учётом границы
            140 + (borderWidth * 2).toInt(), // Высота с учётом границы
            Bitmap.Config.ARGB_8888
        )
        val canvas = Canvas(roundedBitmap)
    
        // Рисуем закругленный фон (с учётом отступов для границы)
        canvas.drawRoundRect(
            RectF(borderWidth, borderWidth, 
                   (markerImage.width + borderWidth).toFloat(), 
                   (markerImage.height + borderWidth).toFloat()), 
            radius, radius, paint
        )
    
        // Рисуем белую границу
        val borderPaint = Paint(ANTI_ALIAS_FLAG)
        borderPaint.color = borderColor
        borderPaint.style = Paint.Style.STROKE // Устанавливаем стиль рисования: только обводка
        borderPaint.strokeWidth = borderWidth // Устанавливаем толщину обводки
    
        // Рисуем границу вокруг закругленного изображения
        canvas.drawRoundRect(
            RectF(borderWidth, borderWidth, 
                   (markerImage.width + borderWidth).toFloat(), 
                   (markerImage.height + borderWidth).toFloat()), 
            radius, radius, borderPaint
        )
    
        return roundedBitmap
    }

    fun setMarkerMapObject(obj: MapObject?) {
        rnMapObject = obj as PlacemarkMapObject?
        rnMapObject!!.addTapListener(this)
        updateMarker()
    }

    fun setChildView(view: View?) {
        if (view == null) {
            _childView!!.removeOnLayoutChangeListener(childLayoutListener)
            _childView = null
            updateMarker()
            return
        }
        _childView = view
        _childView!!.addOnLayoutChangeListener(childLayoutListener)
    }

    fun addChildView(view: View, index: Int) {
        childs.add(index, view)
        setChildView(childs[0])
    }

    fun removeChildView(index: Int) {
        childs.removeAt(index)
        setChildView(if (childs.size > 0) childs[0] else null)
    }

    fun moveAnimationLoop(lat: Double, lon: Double) {
        (rnMapObject as PlacemarkMapObject).geometry = Point(lat, lon)
    }

    fun rotateAnimationLoop(delta: Float) {
        (rnMapObject as PlacemarkMapObject).direction = delta
    }

    fun animatedMoveTo(point: Point, duration: Float) {
        val p = (rnMapObject as PlacemarkMapObject).geometry
        val startLat = p.latitude
        val startLon = p.longitude
        val deltaLat = point.latitude - startLat
        val deltaLon = point.longitude - startLon
        val valueAnimator = ValueAnimator.ofFloat(0f, 1f)
        valueAnimator.setDuration(duration.toLong())
        valueAnimator.interpolator = LinearInterpolator()
        valueAnimator.addUpdateListener { animation ->
            try {
                val v = animation.animatedFraction
                moveAnimationLoop(startLat + v * deltaLat, startLon + v * deltaLon)
            } catch (ex: Exception) {
                // I don't care atm..
            }
        }
        valueAnimator.start()
    }

    fun animatedRotateTo(angle: Float, duration: Float) {
        val placemark = (rnMapObject as PlacemarkMapObject)
        val startDirection = placemark.direction
        val delta = angle - placemark.direction
        val valueAnimator = ValueAnimator.ofFloat(0f, 1f)
        valueAnimator.setDuration(duration.toLong())
        valueAnimator.interpolator = LinearInterpolator()
        valueAnimator.addUpdateListener { animation ->
            try {
                val v = animation.animatedFraction
                rotateAnimationLoop(startDirection + v * delta)
            } catch (ex: Exception) {
                // I don't care atm..
            }
        }
        valueAnimator.start()
    }

    override fun onMapObjectTap(mapObject: MapObject, point: Point): Boolean {
        val e = Arguments.createMap()
        (context as ReactContext).getJSModule(RCTEventEmitter::class.java).receiveEvent(
            id, "onPress", e
        )

        return handled
    }
}
